### DISCUSSION TITLE
Mental model for context based toasts

### DISCUSSION THREAD

**Discussion by t-blackwell** [remix-run/remix#1840] [Q&A] [answered]
I've just started porting my website over to Remix and I've been really impressed with what I've seen so far. The amount of state management/effect code I've been able to strip out is incredible! One stumbling block that I've run into is how toast messages should work in the new Remix world. Previously I would have used a context and handled mutations manually, something like this...

```js
const { showAlert } = useSnackbar();

const handleClick = () => {
  fetch('/do-a-thing', { method: 'POST'})
  .then((response) => response.json())
  .then((data) => {
    showAlert('All good!', 'success');
  })
  .catch((error) => {
    showAlert('Sad times :(', 'error');
  });
}
```

Now that the mutation happens via an action i.e. no explicit event, I don't have a mental model of how I can fire toast messages. I've tried the code below but I can't get it to work...

```js
const actionData = useActionData<ActionData>();
const { showAlert } = useSnackbar();
const transition = useTransition();

// if the formError is populated, the toast shows incorrectly
// when you redirect away from current route
React.useEffect(() => {
  if (actionData?.formError !== undefined && transition.submission) {
    showAlert(actionData.formError, 'error');
  }
}, [actionData?.formError, transition.submission, showAlert]);
```
What am I missing? I've put together a [simple repo](https://github.com/t-blackwell/remix-toast-test) that shows the inner workings of the context etc. in case anyone wants to spin it up.

Any help would be greatly appreciated.

Thanks,
Tom

**Comment by ccssmnn** [comment, marked as answer]
Whenever you submit a form, remix will refetch all loaders to revalidate the data. This keeps all the UI data in your nested components fresh. The flow is:

1. Form is submitted
2. Action has finished with redirect on success and response with an error 
3. all Loaders are called to refresh your data

In your case, you will set error/success in the action and want to know this in a loader. You can use [sessions](https://remix.run/docs/en/v1/api/remix#using-sessions) for this. Store the information in a [sessionCookie](https://remix.run/docs/en/v1/api/remix#createcookiesessionstorage) and read it in a loader / route to render your toast.

**Reply by t-blackwell** [reply]
Thanks for the link, @ccssmnn - that looks like what I'm after. It does feel like overkill to have to put the toast messages into cookies, well to me at least, but do I like the `session.flash` makes it that little bit more transient.

**Comment by kwiat1990** [comment]
Do I get it correctly that I need to somehow use `session.flash` to manage success message after submitting a form? Right know I'm trying to acomplish this task and it turned out to be cumbersome. On the following screenshot you can see states of `useTransition` and `useActionData` in my form component:
<img width="672" alt="Screenshot 2022-06-20 at 21 30 21" src="https://user-images.githubusercontent.com/19494075/174668300-63d62c17-0c13-472b-aeee-cbe65d37e286.png">
Yellow - on initial render
Blue - after submitting and successful response from the server
Green - on every following render (component is toggled in the DOM).

After step 2 (successful form submission) I show a message box instead of the form and it stays in the view during following renders.

**Reply by ccssmnn** [reply]
You don't need a cookie to display a success message for a form submission. You can return an object containing success or error from your action and read from that in your action data.

The original question was seeking for a way to avoid `useContext` to show a toast that lives on one location in the app based on an action anywhere in the app. For this, you can use a cookie, as you might not have access to action data there.

**Reply by kwiat1990** [reply]
@ccssmnn that’s kind of what I’ve tried to do and failed successfully 😉 because after initial form submission the `useActionData` hooks returns data from the server response. The problem is, that if I re-mount a component with the form I still see the message box instead the form because `useActionData` returns previously submitted data. 

I’ve see in docs that `useFetcher` hook has a `type === “done”` but normal forms used with `useTransition` hook don’t have it.

**Reply by ccssmnn** [reply]
Can you provide a codesandbox or stackblitz project reproducing your issue?

**Reply by kwiat1990** [reply]
I have solved my problem using `transition.type === "actionReload"` to check if a form was submitted and data returned from the server.
